tools/checklocks/state.go

```diff
@@ -187,6 +187,35 @@ func (l *lockState) isCompatible(other *lockState) bool {
 	return l.isSubset(other) && other.isSubset(l)
 }
 
+// join returns a new lockState of the intersection of 2 states.
+func (l *lockState) join(other *lockState) *lockState {
+	rls := l.fork()
+
+	for addr, val := range rls.stored {
+		if oval, ok := other.stored[addr]; !ok || val != oval {
+			rls.modify()
+			delete(rls.stored, addr)
+		}
+	}
+
+	if l.isCompatible(other) {
+		return rls
+	}
+
+	for _, m := range rls.lockedMutexes {
```

Comment:
You may want to organize the intersection a bit differently. First to make it O(n) comparisons. Second, I think the current approach can only output 1 element, m.

```
present := make(map[string]bool, len(other.lockedMutexes))
for _, om := range other.lockedMutexes {
  present[om] = true
}
rls.modify() // my understanding is that an intersection must be different given the compat check above.
var w int
for r, n := 0, len(rls.lockedMutexes); i < n; i++ {
  if m := rls[i]; present[m] {
    rls.lockedMutexes[w] = m
    w++
  }
}
rls.lockedMutexes = rls.lockedMutexes[:w]
```
This is a bit complicated as it preserves order and reuses the existing allocation. There are simpler alternatives if this is not needed.

Reply:
Thanks! I think this is a lot better.